### PR TITLE
[FIXED] OCSP: parse cert.Leaf if not set

### PR DESCRIPTION
**Pull request by kozlovic**
When trying to update NATS Streaming server dependencies with
latest NATS Server, I noticed that a TLS test was failing and
this was because the TLS configuration was manually set like this:

```
o := DefaultTestOptions
o.HTTPHost = "127.0.0.1"
o.HTTPSPort = -1
o.TLSConfig = &tls.Config{ServerName: "localhost"}
cert, err := tls.LoadX509KeyPair("configs/certs/server-cert.pem", "configs/certs/server-key.pem")
if err != nil {
    t.Fatalf("Got error reading certificates: %s", err)
}
o.TLSConfig.Certificates = []tls.Certificate{cert}
```
Notice how the `cert.Leaf` is not parsed. This cause the NATS Server
OCSP code to fail when hasOCSPStatusRequest() is invoked with
a `nil` pointer.

My first approach was to add a `nil` check in hasOCSPStatusRequest()
and return `false` in that case.
But then I thought that maybe the correct approach is to parse the
leaf it it is not done in the provided TLS config?

It could be simply a case of fixing the test that I have in
NATS Streaming server repo, but a quick check in this repo's own
dependencies show that not setting the Leaf is something that may
happen in some cases. For instance here is how the Postgres library
build the certs: https://github.com/lib/pq/blob/caa87158f59995a15e2057d80d586391f51333f2/ssl.go#L133

As you can see, the leaf is not parsed here, so I am not sure if
having Leaf nil is valid or not.

The go doc regarding Leaf says:
```
// Leaf is the parsed form of the leaf certificate, which may be initialized
// using x509.ParseCertificate to reduce per-handshake processing. If nil,
// the leaf certificate will be parsed as needed.
Leaf *x509.Certificate
```
This is the last statement that made me chose the current approach of
parsing it if detected as `nil` instead of just ignoring a nil cert.

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

/cc @nats-io/core

### CODE REVIEW

**Review by wallyqs — APPROVED** [review]
LGTM, I think it makes sense to help with the use case of embedded NATS Server.

**Review by variadico — APPROVED** [review]

**Comment by kozlovic** [comment]
@philpennock Would like your opinion on this one, if you don't mind.

File: server/ocsp.go
Comment (philpennock) [inline server/ocsp.go:337]: I'd appreciate a comment here noting the anticipated scenarios where this applies, so that future maintainers know.  "This is normally non-nil, but can still be nil here when in tests or in some embedded scenarios."

There's just too many twisty corner-cases in TLS handling for it to be reasonable to expect all future maintainers to know which logic branches matter when, or "why we bother with this".

File: test/ocsp_test.go
Comment (philpennock) [inline test/ocsp_test.go:2442]: Isn't this testing an implicit config, not a manual config?

File: server/ocsp.go
Comment (kozlovic) [reply server/ocsp.go:339]: Added

File: test/ocsp_test.go
Comment (kozlovic) [reply test/ocsp_test.go]: Actually renamed to `TestOCSPTLSConfigNoLeafSet`

**Comment by kozlovic** [comment]
@philpennock I believe I have addressed your comments. Will merge later today unless you say otherwise.

**Pull request merged**
